### PR TITLE
nConnect client connects to multi servers

### PR DESCRIPTION
**Pull request by billfort**
1. Config multi servers' `remoteAdminAddr`;
2. Query servers' local IP for routing;
3. Set up nkn tunnel for each server;
4. Route data to each server by the server's local IP;
5. Route other data to the default tunnel if the IP or target is not in the routing table.

### CODE REVIEW

File: config.client.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+    "Client": true,
+    "Server": false,
+    "identifier": "alice",
+    "seed": "",
+    "remoteAdminAddr": ["nConnect.bob.7cafe0ae02789f8eb6b293e46b0ac5cf8f92f73042199c8161e5b5f90b13dcb5"],
```
[the remark this answers is not in the document]
Comment (billfort) [reply config.client.json:6]: Yes, modified.

File: README.md [hunk 2]
```diff
@@ -256,6 +255,35 @@ docker run --rm -it --net=host -v ${PWD}:/nConnect/data nknorg/nconnect
 
 followed by the command line argument you want to add.
 
+## nConnect Client Connects to Multi Servers
+
+Now nConnect client can connect to multi servers. You can edit `config.json` to add multi servers admin address:
```
Comment (yilunzhang) [inline README.md:260]: Should we also add examples for cli argument?

Comment (billfort) [reply README.md:260]: Yes. example is added.

File: README.md [hunk 3]
```diff
@@ -256,6 +255,46 @@ docker run --rm -it --net=host -v ${PWD}:/nConnect/data nknorg/nconnect
...
+
+After config multi `remoteAdminAddr`, the nConnect client will add routing infomation to each Servers' local IP. So you can access all the servers by their local IP address.
+
+Specifically, the first  item in the `remoteAdminAddr` will become the default server which will get the forwarded data which targets are beyond all these servers' local IP address. Such as access to website by domain, or some other applications.
+
+You can use command argument to connects to multi servers too. argument `-a` witch address separated by comma "," to pass multi servers addresses:
+
+```
+
+$ nConnect -c -a "server-address1, server-address2, server-address3"
```
Comment (yilunzhang) [inline README.md:287]: Should it be `-a addr1 -a addr2 -a addr3`?

File: nconnect.go [hunk 4]
```diff
@@ -311,29 +316,38 @@ func (nc *nconnect) StartClient() error {
 
 	remoteTunnelAddr := nc.opts.RemoteTunnelAddr
```
Comment (yilunzhang) [inline nconnect.go:317]: `remoteTunnelAddr` is now a pointer to original opts. It's better to make a copy

File: nconnect.go [hunk 5]
```diff
@@ -311,29 +316,38 @@ func (nc *nconnect) StartClient() error {
...
-			if len(remoteInfo.LocalIP.Ipv4) > 0 {
-				vpnRoutes = make([]string, 0, len(remoteInfo.LocalIP.Ipv4))
-				for _, ip := range remoteInfo.LocalIP.Ipv4 {
-					if ip == nc.opts.TunAddr || ip == nc.opts.TunGateway {
-						log.Printf("Skipping server's local IP %s in routes", ip)
-						continue
+			for _, remoteAdminAddr := range nc.opts.RemoteAdminAddr {
+				remoteInfo, err := nc.getRemoteInfo(remoteAdminAddr)
+				if err != nil {
+					return err
```
Comment (yilunzhang) [inline nconnect.go:339]: The logic here is inconsistent with L323.
In L323 (get remote info), failure is printed and remote addr skipped.
Here any failure will cause the app to stop.

File: README.md (hunk 3 above)
Comment (billfort) [reply README.md]: Yes, modified.

File: nconnect.go [hunk 6]
```diff
@@ -356,22 +370,35 @@ func (nc *nconnect) StartClient() error {
...
+		from = append(from, ssAddr)
+		to = append(to, remote)
+
+		if remoteInfo, ok := nc.remoteInfoByTunnel[remote]; ok {
+			for _, addr := range remoteInfo.LocalIP.Ipv4 {
+				nc.ssConfig.Target2Client[addr] = ssAddr
+			}
+		}
+		if index == 0 { // the first config is the default client
+			nc.ssConfig.DefaultClient = ssAddr
```
Comment (yilunzhang) [inline nconnect.go:392]: I would suggest we move all ssconfig updates (L373, L392, L400) to one place

File: nconnect.go [hunk 7]
```diff
@@ -356,22 +370,35 @@ func (nc *nconnect) StartClient() error {
...
+		}
+	}
+	tunnels, err := tunnel.NewTunnels(nc.account, nc.opts.Identifier, from, to, nc.opts.Tuna, nc.tunnelConfig)
 	if err != nil {
 		return err
 	}
-	nc.tunnel = tunnel
+	nc.tunnels = tunnels
+	nc.ssConfig.Client = from[0]
```
Comment (yilunzhang) [inline nconnect.go:400]: Is `ssConfig.Client` still used?

File: nconnect.go [hunk 8]
```diff
@@ -514,13 +541,14 @@ func (nc *nconnect) startSSAndTunnel() {
...
-		if err != nil {
-			log.Fatal(err)
-		}
-		os.Exit(0)
-	}()
+	for _, t := range nc.tunnels {
+		go func(t *tunnel.Tunnel) {
+			err := t.Start()
+			if err != nil {
+				log.Printf("tunnel %v start err %v", t.Addr(), err)
```
Comment (yilunzhang) [inline nconnect.go:548]: I think we should call os.exit like before when any tunnel exits

File: ss/multiconn.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package ss
+
+import (
+	"strings"
+	"sync"
+)
+
+var routes struct {
+	sync.RWMutex
+	Target2Client map[string]string // map target ip to local tunnel port
```
Comment (yilunzhang) [inline ss/multiconn.go:10]: I would suggest we change `2` to `To` follow the same naming convention.

File: ss/multiconn.go [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+var routes struct {
+	sync.RWMutex
+	Target2Client map[string]string // map target ip to local tunnel port
+	DefaultClient string            // the default client for the targets are not in Target2Client map
+}
+
+func getClient(target string) string {
+	tgtIp := strings.Split(target, ":")
```
Comment (yilunzhang) [inline ss/multiconn.go:15]: Is it possible that multiple targets share the same IP address?

File: nconnect.go (hunk 5 above)
Comment (billfort) [reply nconnect.go:339]: Yes, modified.

File: nconnect.go (hunk 7 above)
Comment (billfort) [reply nconnect.go:400]: It is kept just for compatible with ss. 
`
if flags.Client != "" { // client mode
`

File: nconnect.go (hunk 8 above)
Comment (billfort) [reply nconnect.go:548]: Yes, os.exit is added.

File: ss/multiconn.go (hunk 9 above)
Comment (billfort) [reply ss/multiconn.go:10]: Yes, modified.

File: ss/multiconn.go (hunk 10 above)
Comment (billfort) [reply ss/multiconn.go:15]: ss use IP and port to set up connections. Currently it seems difficult to distinguish same IP addresses at ss level although we can distinguish them when `getRemoteInfo` .

File: nconnect.go (hunk 8 above)
Comment (yilunzhang) [reply nconnect.go:546]: The logic is different from the old one, please double check

Comment (billfort) [reply nconnect.go:546]: Yes, modified.

**Review by yilunzhang — APPROVED** [review]
LGTM

**Pull request merged**